### ISSUE TITLE
Debugger only works once after a reboot

### ISSUE THREAD

**Issue by martin-gorner** [go-lang-plugin-org/go-lang-idea-plugin#2360]
- Plugin version (or commit hash): 0.9.1135
- IDE name and version: IntelliJ IDEA 141.3056
- Java version: 1.8.0_31
- OS name and version: OSX 10.11.3

- What are you trying to do?
Debug my program

- What would you expect to happen?
Debugger launching and stopping on my breakpoint

- What happens?
Error message:
/usr/local/go/bin/go build -o "/private/var/folders/85/ty4x5d094flcyvw787q90rr8004_2p/T/Build gohello.go and run2go" -gcflags "-N -l" "/Users/mgorner/Google Drive/workspace/golang/src/gohello/gohello.go"
"/Users/mgorner/Library/Application Support/IntelliJIdea14/Go/lib/dlv/mac/dlv" --listen=localhost:49963 --headless=true exec "/private/var/folders/85/ty4x5d094flcyvw787q90rr8004_2p/T/Build gohello.go and run2go" --
could not launch process: Could not determine version number: could not read string len could not read memory

When I reboot my mac and retry, I get a dialog asking me for my password so as to allow the debugger to take control of the process. The debugger works once after that. The next time I launch it, I have the error message reported above.

**Comment by aren55555** [comment]
I get the same error message (`could not launch process: Unreadable version number: could not read string len could not read memory`) however the dlv debugger works intermittently rather than just once. I haven't been able to decipher what may be the cause of the intermittent successful runs.

**Comment by dlsniper** [comment]
@aren55555 can you please share more information? How do you run the debug process? Which steps? How does your Go Application configuration looks like?

**Comment by pkorobeinikov** [comment]
Have the same problem too.

When debugger succesfully attached, these log messages appears in Idea's log: 

```
2016-03-13 12:16:27,929 [1999021]   INFO -           #com.goide.dlv.DlvVm - OUT: {"method":"RPCServer.CreateBreakpoint","params":[{"file":"/Users/pkorobeinikov/Workspace/golang/src/github.com/pkorobeinikov/beego-mini-example/hello.go","line":20}],"id":1} 
2016-03-13 12:16:27,930 [1999022]   INFO -           #com.goide.dlv.DlvVm - OUT: {"method":"RPCServer.CreateBreakpoint","params":[{"file":"/Users/pkorobeinikov/Workspace/golang/src/github.com/pkorobeinikov/beego-mini-example/hello.go","line":13}],"id":2} 
2016-03-13 12:16:27,930 [1999022]   INFO -           #com.goide.dlv.DlvVm - OUT: {"method":"RPCServer.CreateBreakpoint","params":[{"file":"/Users/pkorobeinikov/Workspace/golang/src/github.com/pkorobeinikov/beego-mini-example/hello.go","line":18}],"id":3} 
2016-03-13 12:16:27,931 [1999023]   INFO -           #com.goide.dlv.DlvVm - OUT: {"method":"RPCServer.CreateBreakpoint","params":[{"file":"/Users/pkorobeinikov/Workspace/golang/src/github.com/pkorobeinikov/beego-mini-example/hello.go","line":17}],"id":4} 
2016-03-13 12:16:27,931 [1999023]   INFO -           #com.goide.dlv.DlvVm - OUT: {"method":"RPCServer.Command","params":[{"Name":"continue"}],"id":5} 
2016-03-13 12:16:28,148 [1999240]   INFO -           #com.goide.dlv.DlvVm - IN: {"id":1,"result":{"id":1,"name":"","addr":8612,"file":"/Users/pkorobeinikov/Workspace/golang/src/github.com/pkorobeinikov/beego-mini-example/hello.go","line":20,"functionName":"main.main","Cond":"","continue":false,"stacktrace":0,"goroutine":false,"hitCount":{},"totalHitCount":0},"error":null} 
2016-03-13 12:16:28,460 [1999552]   INFO -           #com.goide.dlv.DlvVm - IN: {"id":5,"result":{"currentThread":{"id":12547,"pc":8612,"file":"/Users/pkorobeinikov/Workspace/golang/src/github.com/pkorobeinikov/beego-mini-example/hello.go","line":20,"function":{"name":"main.main","value":8336,"type":84,"goType":0,"args":null,"locals":null},"goroutineID":1,"breakPoint":{"id":1,"name":"","addr":8612,"file":"/Users/pkorobeinikov/Workspace/golang/src/github.com/pkorobeinikov/beego-mini-example/hello.go","line":20,"functionName":"main.main","Cond":"","continue":false,"stacktrace":0,"goroutine":false,"hitCount":{"1":1},"totalHitCount":1},"breakPointInfo":{}},"currentGoroutine":{"id":1,"currentLoc":{"pc":3280293,"file":"/usr/local/Cellar/go/1.6/libexec/src/syscall/asm_darwin_amd64.s","line":17,"function":{"name":"syscall.Syscall","value":3280288,"type":84,"goType":0,"args":null,"locals":null}},"userCurrentLoc":{"pc":8612,"file":"/Users/pkorobeinikov/Workspace/golang/src/github.com/pkorobeinikov/beego-mini-example/hello.go","line":20,"function":{"name":"main.main","value":8336,"type":84,"goType":0,"args":null,"locals":null}},"goStatementLoc":{"pc":382129,"file":"/usr/local/Cellar/go/1.6/libexec/src/runtime/asm_amd64.s","line":145,"function":{"name":"runtime.rt0_go","value":381808,"type":84,"goType":0,"args":null,"locals":null}}},"Threads":[{"id":12547,"pc":8612,"file":"/Users/pkorobeinikov/Workspace/golang/src/github.com/pkorobeinikov/beego-mini-example/hello.go","line":20,"function":{"name":"main.main","value":8336,"type":84,"goType":0,"args":null,"locals":null},"goroutineID":1,"breakPoint":{"id":1,"name":"","addr":8612,"file":"/Users/pkorobeinikov/Workspace/golang/src/github.com/pkorobeinikov/beego-mini-example/hello.go","line":20,"functionName":"main.main","Cond":"","continue":false,"stacktrace":0,"goroutine":false,"hitCount":{"1":1},"totalHitCount":1},"breakPointInfo":{}},{"id":13575,"pc":397324,"file":"/usr/local/Cellar/go/1.6/libexec/src/runtime/sys_darwin_amd64.s","line":289,"function":{"name":"runtime.usleep","value":397280,"type":84,"goType":0,"args":null,"locals":null},"goroutineID":0,"breakPointInfo":null},{"id":13827,"pc":397627,"file":"/usr/local/Cellar/go/1.6/libexec/src/runtime/sys_darwin_amd64.s","line":411,"function":{"name":"runtime.mach_semaphore_wait","value":397616,"type":84,"goType":0,"args":null,"locals":null},"goroutineID":0,"breakPointInfo":null},{"id":14083,"pc":397627,"file":"/usr/local/Cellar/go/1.6/libexec/src/runtime/sys_darwin_amd64.s","line":411,"function":{"name":"runtime.mach_semaphore_wait","value":397616,"type":84,"goType":0,"args":null,"locals":null},"goroutineID":0,"breakPointInfo":null},{"id":14339,"pc":397627,"file":"/usr/local/Cellar/go/1.6/libexec/src/runtime/sys_darwin_amd64.s","line":411,"function":{"name":"runtime.mach_semaphore_wait","value":397616,"type":84,"goType":0,"args":null,"locals":null},"goroutineID":0,"breakPointInfo":null}],"exited":false,"exitStatus":0},"error":null} 
2016-03-13 12:16:28,462 [1999554]   INFO -           #com.goide.dlv.DlvVm - OUT: {"method":"RPCServer.StacktraceGoroutine","params":[{"Id":-1,"Depth":100}],"id":6} 
2016-03-13 12:16:28,462 [1999554]   INFO -           #com.goide.dlv.DlvVm - IN: {"id":2,"result":{"id":2,"name":"","addr":8275,"file":"/Users/pkorobeinikov/Workspace/golang/src/github.com/pkorobeinikov/beego-mini-example/hello.go","line":13,"functionName":"main.(*MainController).Get","Cond":"","continue":false,"stacktrace":0,"goroutine":false,"hitCount":{},"totalHitCount":0},"error":null} 
2016-03-13 12:16:28,463 [1999555]   INFO -           #com.goide.dlv.DlvVm - IN: {"id":3,"result":{"id":3,"name":"","addr":8376,"file":"/Users/pkorobeinikov/Workspace/golang/src/github.com/pkorobeinikov/beego-mini-example/hello.go","line":18,"functionName":"main.main","Cond":"","continue":false,"stacktrace":0,"goroutine":false,"hitCount":{},"totalHitCount":0},"error":null} 
2016-03-13 12:16:28,464 [1999556]   INFO -           #com.goide.dlv.DlvVm - IN: {"id":4,"result":{"id":4,"name":"","addr":8367,"file":"/Users/pkorobeinikov/Workspace/golang/src/github.com/pkorobeinikov/beego-mini-example/hello.go","line":17,"functionName":"main.main","Cond":"","continue":false,"stacktrace":0,"goroutine":false,"hitCount":{},"totalHitCount":0},"error":null} 
2016-03-13 12:16:28,464 [1999556]   INFO -           #com.goide.dlv.DlvVm - IN: {"id":6,"result":[{"pc":8612,"file":"/Users/pkorobeinikov/Workspace/golang/src/github.com/pkorobeinikov/beego-mini-example/hello.go","line":20,"function":{"name":"main.main","value":8336,"type":84,"goType":0,"args":null,"locals":null},"Locals":null,"Arguments":null},{"pc":199211,"file":"/usr/local/Cellar/go/1.6/libexec/src/runtime/proc.go","line":188,"function":{"name":"runtime.main","value":198528,"type":84,"goType":0,"args":null,"locals":null},"Locals":null,"Arguments":null},{"pc":393440,"file":"/usr/local/Cellar/go/1.6/libexec/src/runtime/asm_amd64.s","line":1998,"function":{"name":"runtime.goexit","value":393440,"type":84,"goType":0,"args":null,"locals":null},"Locals":null,"Arguments":null}],"error":null} 
2016-03-13 12:16:28,477 [1999569]   INFO -           #com.goide.dlv.DlvVm - OUT: {"method":"RPCServer.ListLocalVars","params":[{"GoroutineID":-1,"Frame":0}],"id":7} 
2016-03-13 12:16:29,471 [2000563]   INFO -           #com.goide.dlv.DlvVm - IN: {"id":7,"result":[{"name":"a","addr":859530632912,"onlyAddr":false,"type":"int","realType":"int","kind":2,"value":"1","len":0,"cap":0,"children":[],"unreadable":""}],"error":null} 
2016-03-13 12:16:29,472 [2000564]   INFO -           #com.goide.dlv.DlvVm - OUT: {"method":"RPCServer.ListFunctionArgs","params":[{"GoroutineID":-1,"Frame":0}],"id":8} 
2016-03-13 12:16:29,473 [2000565]   INFO -           #com.goide.dlv.DlvVm - IN: {"id":8,"result":[],"error":null} 
2016-03-13 12:17:26,754 [2057846]   INFO - #com.goide.dlv.DlvDebugProcess - Connection closed 
org.jetbrains.concurrency.Promise$MessageError: Connection closed
```

When debugger fails to attach, only these messages are logged:

```
2016-03-13 12:17:31,396 [2062488]   INFO -           #com.goide.dlv.DlvVm - OUT: {"method":"RPCServer.CreateBreakpoint","params":[{"file":"/Users/pkorobeinikov/Workspace/golang/src/github.com/pkorobeinikov/beego-mini-example/hello.go","line":20}],"id":1} 
2016-03-13 12:17:31,396 [2062488]   INFO -           #com.goide.dlv.DlvVm - OUT: {"method":"RPCServer.CreateBreakpoint","params":[{"file":"/Users/pkorobeinikov/Workspace/golang/src/github.com/pkorobeinikov/beego-mini-example/hello.go","line":13}],"id":2} 
2016-03-13 12:17:31,397 [2062489]   INFO -           #com.goide.dlv.DlvVm - OUT: {"method":"RPCServer.CreateBreakpoint","params":[{"file":"/Users/pkorobeinikov/Workspace/golang/src/github.com/pkorobeinikov/beego-mini-example/hello.go","line":18}],"id":3} 
2016-03-13 12:17:31,397 [2062489]   INFO -           #com.goide.dlv.DlvVm - OUT: {"method":"RPCServer.CreateBreakpoint","params":[{"file":"/Users/pkorobeinikov/Workspace/golang/src/github.com/pkorobeinikov/beego-mini-example/hello.go","line":17}],"id":4} 
2016-03-13 12:17:31,397 [2062489]   INFO -           #com.goide.dlv.DlvVm - OUT: {"method":"RPCServer.Command","params":[{"Name":"continue"}],"id":5} 
```

No more information nor in debugger console neither in event log window.

**Comment by dlsniper** [comment]
@pkorobeinikov can you please upload the sample code to github.com/pkorobeinikov/beego-mini-example and provide the full log for this? And steps to reproduce this instead of reading the log would make this work faster. Also, which Go version do you have installed? Thank you.

**Comment by pkorobeinikov** [comment]
@dlsniper https://github.com/pkorobeinikov/beego-mini-example

**Comment by dlsniper** [comment]
@pkorobeinikov Thank you.

**Comment by aren55555** [comment]
I am on OSX 10.11.3, running a homebrew built go 1.6 with InteliJ IDEA 15.0.4 and version 0.10.1164 of this plugin. I was using the simplest Go program to just see if the debugging worked:

```go
package main

import (
	"fmt"
)

func main() {
	fmt.Println("Hello World")
}
```

**Comment by fgblomqvist** [comment]
This is the case for me as well. I am running Ubuntu 15.10 and IDEA 2016.1.

**Comment by ngoossens** [comment]
Also experiencing this issue. Same setup as @aren55555 but with the latest nightly 0.10.1199.

**Comment by dlsniper** [comment]
@fgblomqvist I can't replicate it on Ubuntu, same config. Please share more details.
As for OS X I don't have access to such an OS so.. sorry :)

**Comment by aren55555** [comment]
@dlsniper I've also tried with IDEA 2016.1 and the nightly plugin; I get a "Connection refused".

Is there some way I can provide you with more verbose output to help you understand what is going wrong where?

**Comment by dlsniper** [comment]
I've noticed something interesting with the project @pkorobeinikov sent as a sample and 0.11.1197 plugin version.

It seems that at some point after sending the bulk requests to delve, the last command we send is actually the first one that gets returned with an `"exited":true` value. I wonder if the "premature" exit signal happens because of a bug in delve. I can't replicate it manually so I'll ping @derekparker for this, maybe he has some ideas on how to replicate what we do (see the logs below).

The plugin doesn't yet understand the `"exited":true` case, as such, it still tries to run the debugger. @ignatov I think this should be handled by stopping the debugging session without sending any further commands to delve (related issue: https://github.com/go-lang-plugin-org/go-lang-idea-plugin/issues/2399)

```
2016-03-17 22:34:26,563 [2585642]   INFO - lij.diagnostic.DebugLogManager - Set DEBUG for the following categories: #com.goide.dlv.* 
2016-03-17 22:34:55,706 [2614785]   INFO -           #com.goide.dlv.DlvVm - OUT: {"method":"RPCServer.CreateBreakpoint","params":[{"file":"/home/florin/golang/src/github.com/dlsniper/u/demo/demo2360/dbg.go","line":13}],"id":1} 
2016-03-17 22:34:55,706 [2614785]   INFO -           #com.goide.dlv.DlvVm - OUT: {"method":"RPCServer.CreateBreakpoint","params":[{"file":"/home/florin/golang/src/github.com/dlsniper/u/demo/demo2360/dbg.go","line":17}],"id":2} 
2016-03-17 22:34:55,707 [2614786]   INFO -           #com.goide.dlv.DlvVm - OUT: {"method":"RPCServer.CreateBreakpoint","params":[{"file":"/home/florin/golang/src/github.com/dlsniper/u/demo/demo2360/dbg.go","line":18}],"id":3} 
2016-03-17 22:34:55,707 [2614786]   INFO -           #com.goide.dlv.DlvVm - OUT: {"method":"RPCServer.CreateBreakpoint","params":[{"file":"/home/florin/golang/src/github.com/dlsniper/u/demo/demo2360/dbg.go","line":20}],"id":4} 
2016-03-17 22:34:55,707 [2614786]   INFO -           #com.goide.dlv.DlvVm - OUT: {"method":"RPCServer.Command","params":[{"Name":"continue"}],"id":5} 
2016-03-17 22:34:55,717 [2614796]   INFO -           #com.goide.dlv.DlvVm - IN: {"id":1,"result":{"id":1,"name":"","addr":4198419,"file":"/home/florin/golang/src/github.com/dlsniper/u/demo/demo2360/dbg.go","line":13,"functionName":"main.(*MainController).Get","Cond":"","continue":false,"stacktrace":0,"goroutine":false,"hitCount":{},"totalHitCount":0},"error":null} 
2016-03-17 22:36:09,421 [2688500]   INFO -           #com.goide.dlv.DlvVm - IN: {"id":5,"result":{"Threads":null,"exited":true,"exitStatus":0},"error":null} 
2016-03-17 22:36:09,422 [2688501]   INFO -           #com.goide.dlv.DlvVm - OUT: {"method":"RPCServer.Detach","params":[true],"id":6} 
2016-03-17 22:36:09,422 [2688501]   INFO -           #com.goide.dlv.DlvVm - IN: {"id":2,"result":null,"error":"no such process"} 
2016-03-17 22:36:09,422 [2688501]   INFO - #com.goide.dlv.DlvDebugProcess - no such process 
org.jetbrains.concurrency.Promise$MessageError: no such process
2016-03-17 22:36:09,422 [2688501]   INFO -           #com.goide.dlv.DlvVm - IN: {"id":3,"result":null,"error":"no such process"} 
2016-03-17 22:36:09,422 [2688501]   INFO - #com.goide.dlv.DlvDebugProcess - no such process 
org.jetbrains.concurrency.Promise$MessageError: no such process
2016-03-17 22:36:09,422 [2688501]   INFO -           #com.goide.dlv.DlvVm - IN: {"id":4,"result":null,"error":"no such process"} 
2016-03-17 22:36:09,422 [2688501]   INFO - #com.goide.dlv.DlvDebugProcess - no such process 
org.jetbrains.concurrency.Promise$MessageError: no such process
2016-03-17 22:36:09,422 [2688501]   WARN - channel.DefaultChannelPipeline - An exceptionCaught() event was fired, and it reached at the tail of the pipeline. It usually means the last handler in the pipeline did not handle the exception. 
java.net.SocketException: Connection reset
	at java.net.SocketInputStream.read(SocketInputStream.java:209)
	at java.net.SocketInputStream.read(SocketInputStream.java:141)
	at io.netty.buffer.UnsafeByteBufUtil.setBytes(UnsafeByteBufUtil.java:472)
	at io.netty.buffer.PooledUnsafeDirectByteBuf.setBytes(PooledUnsafeDirectByteBuf.java:251)
	at io.netty.buffer.AbstractByteBuf.writeBytes(AbstractByteBuf.java:1034)
	at io.netty.channel.oio.OioByteStreamChannel.doReadBytes(OioByteStreamChannel.java:109)
	at io.netty.channel.socket.oio.OioSocketChannel.doReadBytes(OioSocketChannel.java:139)
	at io.netty.channel.oio.AbstractOioByteChannel.doRead(AbstractOioByteChannel.java:130)
	at io.netty.channel.oio.AbstractOioChannel$1.run(AbstractOioChannel.java:38)
	at io.netty.channel.ThreadPerChannelEventLoop.run(ThreadPerChannelEventLoop.java:52)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:742)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
2016-03-17 22:36:09,473 [2688552]   INFO - #com.goide.dlv.DlvDebugProcess - Connection closed 
org.jetbrains.concurrency.Promise$MessageError: Connection closed
```

**Comment by dlsniper** [comment]
> Is there some way I can provide you with more verbose output to help you understand what is going wrong where?

@aren55555 you can provide a code sample which reproduces the issue. Also logs are really helpful in these situations so you could look into ` Help | Show logs in ... ` and supply the logs for the debugging session when this fails (see an example from the above of how they look like). Also, you can try to see what happens when you debug from the command line using delve vs from the IDE and finally you can have a look in your process manager to see if there's any lingering debugging / debugged process around (or something else blocking the port delve is trying to launch on).

**Comment by aren55555** [comment]
@dlsniper could not locate the `Help | Show logs in ...`... I think perhaps the Mac IDEA app does not have that. However via this resource I found that the logs are located in `~/Library/Logs/IntelliJIdea15/idea.log`. I tailed this file and every time the `could not launch process: Unreadable version number: could not read string len could not read memory` error was shown in the app the following log resulted:

```plain
2016-03-17 16:59:43,426 [ 367741]  ERROR - ecution.process.ProcessHandler - null 
java.lang.reflect.InvocationTargetException
	at sun.reflect.GeneratedMethodAccessor100.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at com.intellij.execution.process.ProcessHandler$4.invoke(ProcessHandler.java:214)
	at com.sun.proxy.$Proxy15.processTerminated(Unknown Source)
	at com.intellij.execution.process.ProcessHandler$3.run(ProcessHandler.java:171)
	at com.intellij.execution.process.ProcessHandler$TasksRunner.execute(ProcessHandler.java:247)
	at com.intellij.execution.process.ProcessHandler.notifyTerminated(ProcessHandler.java:153)
	at com.intellij.execution.process.ProcessHandler.notifyProcessTerminated(ProcessHandler.java:149)
	at com.intellij.execution.process.KillableColoredProcessHandler.notifyProcessTerminated(KillableColoredProcessHandler.java:127)
	at com.intellij.execution.process.BaseOSProcessHandler.onOSProcessTerminated(BaseOSProcessHandler.java:150)
	at com.intellij.execution.process.BaseOSProcessHandler$1$1.consume(BaseOSProcessHandler.java:115)
	at com.intellij.execution.process.BaseOSProcessHandler$1$1.consume(BaseOSProcessHandler.java:100)
	at com.intellij.execution.process.ProcessWaitFor$1.run(ProcessWaitFor.java:55)
	at com.intellij.openapi.application.impl.ApplicationImpl$8.run(ApplicationImpl.java:366)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
	at org.jetbrains.ide.PooledThreadExecutor$1$1.run(PooledThreadExecutor.java:55)
Caused by: java.lang.IllegalArgumentException: Argument for @NotNull parameter 't' of com/intellij/util/ObjectUtils._assertNotNull must not be null
	at com.intellij.util.ObjectUtils._assertNotNull(ObjectUtils.java)
	at com.intellij.util.ObjectUtils.assertNotNull(ObjectUtils.java:35)
	at com.goide.dlv.DlvDebugProcess.getProcessor(DlvDebugProcess.java:126)
	at com.goide.dlv.DlvDebugProcess.send(DlvDebugProcess.java:116)
	at com.goide.dlv.DlvDebugProcess.stop(DlvDebugProcess.java:234)
	at com.intellij.xdebugger.XDebugProcess.stopAsync(XDebugProcess.java:140)
	at com.intellij.xdebugger.impl.XDebugSessionImpl.d(XDebugSessionImpl.java:887)
	at com.intellij.xdebugger.impl.XDebugSessionImpl.access$000(XDebugSessionImpl.java:88)
	at com.intellij.xdebugger.impl.XDebugSessionImpl$1.processTerminated(XDebugSessionImpl.java:277)
	... 21 more
2016-03-17 16:59:43,427 [ 367742]  ERROR - ecution.process.ProcessHandler - IntelliJ IDEA 15.0.4  Build #IU-143.2287.1 
2016-03-17 16:59:43,427 [ 367742]  ERROR - ecution.process.ProcessHandler - JDK: 1.8.0_40-release 
2016-03-17 16:59:43,427 [ 367742]  ERROR - ecution.process.ProcessHandler - VM: OpenJDK 64-Bit Server VM 
2016-03-17 16:59:43,427 [ 367742]  ERROR - ecution.process.ProcessHandler - Vendor: JetBrains s.r.o 
2016-03-17 16:59:43,427 [ 367742]  ERROR - ecution.process.ProcessHandler - OS: Mac OS X 
2016-03-17 16:59:43,427 [ 367742]  ERROR - ecution.process.ProcessHandler - Last Action: Debug 

```

**Comment by zolotov** [comment]
@aren55555 please attach whole log (e.g. via gist.github.com), not tail.

Also, Mac IDEA for sure has Show Log action.

<img width="420" alt="screenshot 2016-03-18 03 08 25" src="https://cloud.githubusercontent.com/assets/140920/13864934/d24a6f20-ecb6-11e5-8f68-2f5f8ea7f3cd.png">

**Comment by aren55555** [comment]
@zolotov https://gist.github.com/aren55555/3d020e0189b4e298b6c0

**Comment by derekparker** [comment]
@dlsniper thanks for pinging me on this. There is a known issue on OSX during initialization of the inferior process, that stems from how Delve has to fork/exec the process on OSX. This is being looked into now and should have a fix soon. I've created https://github.com/derekparker/delve/issues/464 to track this.

**Comment by fgblomqvist** [comment]
Is the issue the same on Ubuntu or should I open a different issue for that?
Here's the log:
http://paste.ubuntu.com/15592170/
(I'm starting IDEA, debugging once -- which works -- then debugging a second time, which doesn't work)

**Comment by dlsniper** [comment]
@fgblomqvist can you please get a sample of code that reproduces this issue? There's no way to tell without it if the problem is in delve or in other part. Thanks.

**Comment by fgblomqvist** [comment]
@dlsniper Here is a simple piece of code, very simple: http://paste.ubuntu.com/15592326/
And here is the log when I did the same as above but with this piece of code: http://paste.ubuntu.com/15592341/

**Comment by dlsniper** [comment]
@ignatov until this gets a proper fix in Delve I'm thinking of applying a temporary fix to allow for a command to be sent only after the previous one has returned. What do you think?

**Comment by ohohco** [comment]
Mac can't use debug. Windows is well.

**Comment by dlsniper** [comment]
@ohohco what does that mean? What plugin version do you have on those platforms? How does the code you want to debug looks like? Can you please open a new issues for your problem on Mac?

**Comment by theganyo** [comment]
This seems to match my experience on OS X as well. The debugger works once and then fails to work with breakpoints after that.

**Comment by dlsniper** [comment]
Can you please provide the ide log? I can give you the steps to get it if needed. It might be the same problem or a different one. Thank you.

**Comment by dmsoft** [comment]
I'm the next one confirming the same behaviour. Debugger stops at breakpoints the first time only... Then fails to stop for random period of time. I didn't restart just starts debugging in some time.


`2016-04-13 00:34:23,480 [1142214]   INFO -           #com.goide.dlv.DlvVm - OUT: {"method":"RPCServer.CreateBreakpoint","params":[{"file":"/Users/dm/go/src/taptor/src/taptor.go","line":194}],"id":1} 
2016-04-13 00:34:23,480 [1142214]   INFO -           #com.goide.dlv.DlvVm - OUT: {"method":"RPCServer.Command","params":[{"Name":"continue"}],"id":2} 
2016-04-13 00:34:23,561 [1142295]   INFO -           #com.goide.dlv.DlvVm - IN: {"id":2,"result":{"Threads":null,"exited":true,"exitStatus":0},"error":null} 
2016-04-13 00:34:23,562 [1142296]   INFO -           #com.goide.dlv.DlvVm - OUT: {"method":"RPCServer.Detach","params":[true],"id":3} 
2016-04-13 00:34:23,608 [1142342]   INFO -           #com.goide.dlv.DlvVm - IN: {"id":1,"result":null,"error":"Process 0 has exited with status 0"} 
2016-04-13 00:34:23,609 [1142343]   INFO - #com.goide.dlv.DlvDebugProcess - Process 0 has exited with status 0 
org.jetbrains.concurrency.Promise$MessageError: Process 0 has exited with status 0
2016-04-13 00:34:23,609 [1142343]   INFO -           #com.goide.dlv.DlvVm - IN: {"id":3,"result":0,"error":null} 
2016-04-13 00:34:23,624 [1142358]   INFO - #com.goide.dlv.DlvDebugProcess - Connection closed 
org.jetbrains.concurrency.Promise$MessageError: Connection closed
2016-04-13 00:35:13,410 [1192144]   INFO -           #com.goide.dlv.DlvVm - OUT: {"method":"RPCServer.CreateBreakpoint","params":[{"file":"/Users/dm/go/src/taptor/src/taptor.go","line":194}],"id":1} 
2016-04-13 00:35:13,410 [1192144]   INFO -           #com.goide.dlv.DlvVm - OUT: {"method":"RPCServer.Command","params":[{"Name":"continue"}],"id":2} 
2016-04-13 00:35:13,575 [1192309]   INFO -           #com.goide.dlv.DlvVm - IN: {"id":2,"result":{"Threads":null,"exited":true,"exitStatus":0},"error":null} 
2016-04-13 00:35:13,577 [1192311]   INFO -           #com.goide.dlv.DlvVm - OUT: {"method":"RPCServer.Detach","params":[true],"id":3} 
2016-04-13 00:35:13,621 [1192355]   INFO -           #com.goide.dlv.DlvVm - IN: {"id":1,"result":null,"error":"Process 0 has exited with status 0"} 
2016-04-13 00:35:13,621 [1192355]   INFO - #com.goide.dlv.DlvDebugProcess - Process 0 has exited with status 0 
org.jetbrains.concurrency.Promise$MessageError: Process 0 has exited with status 0
2016-04-13 00:35:13,622 [1192356]   INFO -           #com.goide.dlv.DlvVm - IN: {"id":3,"result":0,"error":null} 
2016-04-13 00:35:13,626 [1192360]   INFO - #com.goide.dlv.DlvDebugProcess - Connection closed 
org.jetbrains.concurrency.Promise$MessageError: Connection closed
2016-04-13 00:35:29,780 [1208514]   INFO - lij.tasks.impl.TaskManagerImpl - Updating issues cache (every 20 min) `

**Comment by dlsniper** [comment]
Thank you for the logs. Please understand that this is an upstream issue as well. Surely someone could help Derek fix this, no?

**Comment by dlsniper** [comment]
@ignatov I've been thinking a bit about the issue. Would it make sense to "serialize" the requests stream into a sequential requests rather than parallel requests? To achieve the least disruption for the user, I think we could say something like this int:
```
// in the send request part
if request == continue | step (in|out|next) then 
  //wait until all in-flight requests have finished
  wait until current response counter = current request id - 1
end
send request

// in the receive request part
get response
increment current response counter
```

I'd be happy to try and provide a patch for this, however I don't know how to do the "wait" part.

What do you think?

**Comment by fane89** [comment]
I have the same issue - mac osx.

Here are my logs from intelliJ:

`2016-04-22 22:48:28,933 [ 353264]   INFO -           #com.goide.dlv.DlvVm - OUT: {"method":"RPCServer.CreateBreakpoint","params":[{"file":"/xxxxxxxx","line":8}],"id":1} 
2016-04-22 22:48:28,933 [ 353264]   INFO -           #com.goide.dlv.DlvVm - OUT: {"method":"RPCServer.Command","params":[{"Name":"continue"}],"id":2} 
2016-04-22 22:48:28,974 [ 353305]   INFO -           #com.goide.dlv.DlvVm - IN: {"id":2,"result":{"Threads":null,"exited":true,"exitStatus":1},"error":null} 
2016-04-22 22:48:28,981 [ 353312]   INFO -           #com.goide.dlv.DlvVm - OUT: {"method":"RPCServer.Detach","params":[true],"id":3} 
2016-04-22 22:48:29,018 [ 353349]   INFO -           #com.goide.dlv.DlvVm - IN: {"id":1,"result":null,"error":"Process 0 has exited with status 0"} 
2016-04-22 22:48:29,018 [ 353349]   INFO - #com.goide.dlv.DlvDebugProcess - Process 0 has exited with status 0 
org.jetbrains.concurrency.Promise$MessageError: Process 0 has exited with status 0
2016-04-22 22:48:29,019 [ 353350]   INFO -           #com.goide.dlv.DlvVm - IN: {"id":3,"result":0,"error":null} 
2016-04-22 22:48:29,056 [ 353387]   INFO - #com.goide.dlv.DlvDebugProcess - Connection closed 
org.jetbrains.concurrency.Promise$MessageError: Connection closed`

**Comment by dlsniper** [comment]
I'm sorry, locking the conversation until this gets fixed or either @ignatov / @zolotov unlocks it. There's a bunch of me too about the same problem which don't lead to any fix either in the plugin or in delve.
I really appreciate the situation people are in but for now it's a known this type of problem is known and it has to be addressed somehow.
If someone wants to help out, either by fixing the issue in delve or in the plugin that's excellent, otherwise this doesn't help too much unfortunately.

**Comment by dlsniper** [comment]
Hi all, there's an experimental fix here: https://github.com/go-lang-plugin-org/go-lang-idea-plugin/pull/2488 I can't guarantee that it will end up in the plugin but if someone that can consistently reproduce the issue test it check it out, that would be awesome. Feel free to ping me on: https://gitter.im/go-lang-plugin-org/go-lang-idea-plugin if you need any help / provide feedback. Thanks.

**Comment by dlsniper** [comment]
I've done more digging last night and this appears to be in a rather strange situation.
IntelliJ reports that we send the requests in order, and every time I've tried to observe how the requests are sent out this looked fine.
As soon as I've removed the breakpoints from the plugin and ran delve with some debug lines to print the requests as soon as they came in, it printed out that the continue request came in before the breakpoints one did.
I've looked on how the platform sends the requests and it appears to be in order but I'm stuck on this one for now, I'll continue investigating this.

**Comment by dlsniper** [comment]
To give an update on this, the problem has been traced in Delve. There's an issue raised in Delve, https://github.com/derekparker/delve/issues/523 and PR https://github.com/derekparker/delve/pull/520 that while it solves the problem at hand, it breaks other functionalities so it can be merged.

Given this problem is on Delve's communication layer and not in the debugger itself or any of the other "hard" parts, I would ask anyone that has knowledge on Go and the time to look into the issue to help out. As soon as this gets fixed in Delve, the debugging in the plugin will work again.

Thank you.

**Comment by dlsniper** [comment]
fixed in #2599
